File: Python/strategy/duck.py
from abc import ABC
from abc import abstractmethod

from strategy.fly_behivor import FlyBehavior
from strategy.fly_behivor import FlyNoWay
from strategy.fly_behivor import FlyWithWings
from strategy.quack_behavior import MuteQuack
from strategy.quack_behavior import Quack
from strategy.quack_behavior import QuackBehavior
from strategy.quack_behavior import Squeak


class Duck(ABC):
    fly_behavior: FlyBehavior
    quack_behavior: QuackBehavior

    @abstractmethod
    def display(self):
        pass

    def swim(self):
        print("물에 둥둥")

    def perform_fly(self):
        self.fly_behavior.fly()

    def perform_quack(self):
        self.quack_behavior.quack()

    def set_fly_behavior(self, fly):
        self.fly_behavior = fly

    def set_quack_behavior(self, quack):
        self.quack_behavior = quack


class MallardDuck(Duck):
    def __init__(self):
        self.fly_behavior = FlyWithWings()
        self.quack_behavior = Quack()

    def display(self):
        print("MallardDuck")


class RedHeadDuck(Duck):
    def __init__(self):
        self.fly_behavior = FlyWithWings()
        self.quack_behavior = Squeak()

    def display(self):
        print("RedHeadDuck")


class RubberDuck(Duck):
    def __init__(self):
        self.fly_behavior = FlyNoWay()
        self.quack_behavior = MuteQuack()

    def display(self):
        print("RubberDuck")

File: Scala/src/main/scala/strategy/Duck.scala
package strategy

trait Duck {

  val flyBehavior: () => Unit
  val quackBehavior: () => Unit

  def display(): Unit

  def swim(): Unit = println("물에 둥둥")

  def performFly(f: () => Unit = flyBehavior): Unit = f()

  def performQuack(f: () => Unit = quackBehavior): Unit = f()

}


class MallardDuck extends Duck {
  override val flyBehavior: () => Unit = FlyBehavior.flyWithWings
  override val quackBehavior: () => Unit = QuackBehavior.quack

  override def display(): Unit = println("MallardDuck")
}


class RedHeadDuck extends Duck {
  override val flyBehavior: () => Unit = FlyBehavior.flyWithWings
  override val quackBehavior: () => Unit = QuackBehavior.squeak

  override def display(): Unit = println("RedHeadDuck")
}

class RubberDuck extends Duck {
  override val flyBehavior: () => Unit = FlyBehavior.flyNoWay
  override val quackBehavior: () => Unit = QuackBehavior.muteQuack

  override def display(): Unit = println("RubberDuck")
}
